src/v/cloud_storage/tests/directory_walker_test.cc

```diff
@@ -0,0 +1,161 @@
+ * Copyright 2021 Vectorized, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/vectorizedio/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "bytes/iobuf.h"
+#include "cloud_storage/recursive_directory_walker.h"
+#include "seastarx.h"
+
+#include <seastar/core/file.hh>
+#include <seastar/core/fstream.hh>
+#include <seastar/core/seastar.hh>
+#include <seastar/core/sleep.hh>
+#include <seastar/testing/test_case.hh>
+#include <seastar/testing/thread_test_case.hh>
+
+#include <boost/test/tools/old/interface.hpp>
+#include <boost/test/unit_test.hpp>
+
+#include <chrono>
+
+using namespace cloud_storage;
+using namespace std::chrono_literals;
+
+SEASTAR_THREAD_TEST_CASE(one_level) {
+    cloud_storage::recursive_directory_walker _walker;
+    const std::filesystem::path target_dir{"target_dir"};
+    const std::filesystem::path file_path1{"target_dir/file1.txt"};
+    const std::filesystem::path file_path2{"target_dir/file2.txt"};
+
+    ss::recursive_touch_directory(target_dir.native()).get();
+
+    auto flags = ss::open_flags::wo | ss::open_flags::create
+                 | ss::open_flags::exclusive;
+    auto file1 = ss::open_file_dma(file_path1.native(), flags).get();
+    file1.close().get();
+
+    ss::sleep(ss::lowres_clock::duration(1s)).get();
+
+    auto file2 = ss::open_file_dma(file_path2.native(), flags).get();
+    file2.close().get();
+
+    auto result = _walker.walk(target_dir.native()).get();
+
+    BOOST_REQUIRE_EQUAL(result.first, 0);
+    BOOST_REQUIRE_EQUAL(result.second.size(), 2);
+    BOOST_REQUIRE_EQUAL(result.second[0].path, file_path1.native());
+    BOOST_REQUIRE_EQUAL(result.second[1].path, file_path2.native());
+}
+
+SEASTAR_THREAD_TEST_CASE(three_levels) {
+    cloud_storage::recursive_directory_walker _walker;
+    const std::filesystem::path target_dir{"target_dir1"};
+    const std::filesystem::path file_path1{"target_dir1/a/file1.txt"};
+    const std::filesystem::path file_path2{"target_dir1/file2.txt"};
+    const std::filesystem::path file_path3{"target_dir1/b/c/file3.txt"};
+
+    ss::recursive_touch_directory(target_dir.native()).get();
+    ss::recursive_touch_directory("target_dir1/a").get();
+    ss::recursive_touch_directory("target_dir1/b/c").get();
+
+    auto flags = ss::open_flags::wo | ss::open_flags::create
+                 | ss::open_flags::exclusive;
+    auto file1 = ss::open_file_dma(file_path1.native(), flags).get();
+    file1.close().get();
+    ss::sleep(ss::lowres_clock::duration(1s)).get();
+
+    auto file2 = ss::open_file_dma(file_path2.native(), flags).get();
+    file2.close().get();
+    ss::sleep(ss::lowres_clock::duration(1s)).get();
+
+    auto file3 = ss::open_file_dma(file_path3.native(), flags).get();
+    file3.close().get();
+    ss::sleep(ss::lowres_clock::duration(1s)).get();
```

Comment:
this could use a loop over a list of paths :)

Reply:
yes, but I like to make the code in unit tests as dummy as possible, this way it's very obvious what's happening